### ISSUE TITLE
After replace `torch.Tensor.__getitem__` with some other implementation, indexing with tensors of one dim behaves unexpectedly.

### ISSUE THREAD

**Issue by ChieloNewctle** [pytorch/pytorch#41849]
## To Reproduce

```
import torch

a = torch.tensor([[1, 2, 3], [4, 5, 6]])
b = torch.tensor([True, True])

print(a[b].shape)

print('__getitem__' in torch.Tensor.__dict__)
f = torch.Tensor.__getitem__
g = lambda *a, **kw: f(*a, **kw)
torch.Tensor.__getitem__ = g
print('__getitem__' in torch.Tensor.__dict__)

print(a[b].shape)

delattr(torch.Tensor, '__getitem__')
print('__getitem__' in torch.Tensor.__dict__)

print(a[b].shape)
```

This code gives:
```
torch.Size([2, 3])
False
True
torch.Size([1, 2, 3])
False
torch.Size([1, 2, 3])
```

## Expected behavior

The code should have outputs like:
```
torch.Size([2, 3])
False
True
torch.Size([2, 3])
False
torch.Size([2, 3])
```

Or it's even okay to have:
```
torch.Size([2, 3])
False
True
torch.Size([1, 2, 3])
False
torch.Size([2, 3])
```

The most annoying problem is, when I undo the modification, indexing is still wrong.

## Environment

```
Collecting environment information...
PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1060 6GB
Nvidia driver version: 450.51.05
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.17.5
[pip3] torch==1.3.1
[pip3] torchvision==0.4.2
[conda] Could not collect
```

## Additional context

If we set `torch.Tensor.__getitem__` with `f`, which is the original `__getitem__` method from `torch._C._TensorBase`, the code behaves well.

**Comment by albanD** [comment]
Don't you just need to put back the `__getitem__` to get back the pytorch behavior?
If you remove our implementation of `__getitem__`, it is expected to not behave properly.

**Comment by ChieloNewctle** [comment]
> Don't you just need to put back the `__getitem__` to get back the pytorch behavior?
> If you remove our implementation of `__getitem__`, it is expected to not behave properly.

`Tensor` doesn't actually have `__getitem__` attribute. But it's true I should put back `__getitem__` for test.

And here is the code:
```
import torch

a = torch.tensor([[1, 2, 3], [4, 5, 6]])
b = torch.tensor([True, True])

print(a[b].shape)

print('__getitem__' in torch.Tensor.__dict__)
f = torch.Tensor.__getitem__
g = lambda *a, **kw: f(*a, **kw)
torch.Tensor.__getitem__ = g
print('__getitem__' in torch.Tensor.__dict__)

print(a[b].shape)

#### MODIFIED HERE ####
setattr(torch.Tensor, '__getitem__', f)    # delattr(torch.Tensor, '__getitem__')
print('__getitem__' in torch.Tensor.__dict__)

print(a[b].shape)
```

And this code outputs:
```
torch.Size([2, 3])
False
True
torch.Size([1, 2, 3])
True
torch.Size([1, 2, 3])
```

It's still not working properly... :)

**Comment by albanD** [comment]
Given that we don't define these functions one by one but by making the Tensor a Sequence object. You might get weird behavior when you try to change one as a plain function. Not sure.

**Comment by ChieloNewctle** [comment]
@albanD I debugged `THPVariable_getitem` parallelly under both normal and abnormal circumstances.

When calling `treatSequenceAsTuple` ([here](https://github.com/pytorch/pytorch/blob/72b0447f8d22413a388e794e122e3056bffbb735/torch/csrc/autograd/python_variable_indexing.cpp#L204)), I find `PySequence_Check` returns `1` instead of `0` after changing `__getitem__` of `Tensor`.

But even if I revert the modification (delete the attribute), the problem still remains. So the problem now is about CPython.

I'll close this issue. Thanks for your help.
